### PR TITLE
Assignment9+10

### PR DESCRIPTION
**Pull request by NicholasJohnsonDev**
Teammate Luis Garcia @TechGlue

### CODE REVIEW

**Comment by sammyboy2000** [comment]
I will review this.

**Comment by sammyboy2000** [comment]
## Instructions

1. Implement `PingProcess`' `public Task<PingResult> RunTaskAsync(string hostNameOrAddress)` ✔
   - First implement `public void RunTaskAsync_Success()` test method to test `PingProcess.RunTaskAsync()` using `"localhost"`. ✔
   - Do NOT use async/await in this implementation. ✔
2. Implement `PingProcess`' `async public Task<PingResult> RunAsync(string hostNameOrAddress)` ✔
   - First implement the `public void RunAsync_UsingTaskReturn_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` **without** using async/await. ✔
   - Also implement the `async public Task RunAsync_UsingTpl_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` but this time **DO** using async/await. ✔
3. Add support for an optional cancellation token to the `PingProcess.RunAsync()` signature. ✔
   Inside the `PingProcess.RunAsync()` invoke the token's `ThrowIfCancellationRequested()` method so an exception is thrown. ✔
   Test that, when cancelled from the test method, the exception thrown is an `AggregateException` ✔ with a `TaskCanceledException` inner exception ✔ using the test methods `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping` ✔and `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException` ✔ respectively.
4. Complete/fix **AND test** `async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default)` which executes ping for and array of hostNameOrAddresses (which can all be "localhost") **in parallel**, adding synchronization if needed. ✔
   NOTE:
      - The order of the items in the stdOutput is irrelevent and expected to be intermingled.
      - StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔
5. Implement **AND test** `public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token)` using `Task.Factory.StartNew()` and invoking `RunProcessInternal` with a `TaskCreation` value of `TaskCreationOptions.LongRunning` and a `TaskScheduler` value of `TaskScheduler.Current`. ❌
   NOTE: This method does **NOT** use `Task.Run`.

## Extra Credit

- Test and implement `PingProcess.RunAsync(System.IProgress<T> progess)` so that you can capture the output as it occurs rather than capturing the output only when the process finishes. ❌

## Fundamentals

- Ensure you enable:
  - nullable reference types is enabled ❌
  - net6 targeted ❌
  - C# 10.0 ❌
  - and enabled .NET analyzers for both projects ❌
- All of the above should be unit tested ❌
- Choose simplicity over complexity ✔

Looking great so far. All you need to do is enable your fundamentals in the csproj files and finish part 5.

**Comment by NicholasJohnsonDev** [comment]
Those "fundementals" are done in the  https://github.com/IntelliTect-Samples/EWU-CSCD371-2022-Winter/blob/64b268c305432d4dcf8d5dd9a9cc99ce5509bd65/Assignment/Directory.Build.props file.

Question 5 was finished and some minor adjustments were made.
Made adjustments per warnings or suppressed them.

File: Assignment/Assignment/PingProcess.cs [hunk 1]
```diff
@@ -13,54 +13,87 @@ public record struct PingResult(int ExitCode, string? StdOutput);
...
-        Task task = null!;
-        await task;
-        throw new NotImplementedException();
+        StartInfo.Arguments = hostNameOrAddress;
+        StringBuilder? stringBuilder = null;
+        void UpdateStdOutput(string? line) =>
+            (stringBuilder ??= new StringBuilder()).AppendLine(line);
+
+        Task<PingResult> task = Task.Run(
+            () =>
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:48]: I would keep this on the line above

File: Assignment/Assignment/PingProcess.cs [hunk 2]
```diff
@@ -13,54 +13,87 @@ public record struct PingResult(int ExitCode, string? StdOutput);
...
+            {
+                Process process = RunProcessInternal(StartInfo, UpdateStdOutput, default, cancellationToken);
+                cancellationToken.ThrowIfCancellationRequested();
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:51]: You may want to do this on the line above

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 3]
```diff
@@ -58,65 +59,90 @@ public void Run_CaptureStdOutput_Success()
...
     [TestMethod]
-#pragma warning disable CS1998 // Remove this
     async public Task RunAsync_UsingTpl_Success()
     {
         // DO use async/await in this test.
-        PingResult result = default;
-
-        // Test Sut.RunAsync("localhost");
+        Task<PingResult> task = Sut.RunAsync("localhost");
+        PingResult result = await task;
```
Comment (mmwoodfo) [inline Assignment/Assignment.Tests/PingProcessTests.cs:88]: You would typically do these on the same line:
```suggestion
        PingResult task = await Sut.RunAsync("localhost");
```

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 4]
```diff
@@ -58,65 +59,90 @@ public void Run_CaptureStdOutput_Success()
...
     [ExpectedException(typeof(TaskCanceledException))]
-    public void RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException()
+    public async Task RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException()
     {
-        // Use exception.Flatten()
+        //Create a new token
+        CancellationTokenSource cancellationTokenSource = new();
+
+        //Cancel Token
+        cancellationTokenSource.Cancel();
```
Comment (mmwoodfo) [inline Assignment/Assignment.Tests/PingProcessTests.cs:119]: You should probably wait to cancel until after you've started RunAsync

File: Assignment/Assignment/PingProcess.cs [hunk 5]
```diff
@@ -13,54 +13,87 @@ public record struct PingResult(int ExitCode, string? StdOutput);
...
+            => compiledString.Trim() + currentString.Result.StdOutput).Trim());
+
+        // int? lineCount = stringBuilder?.ToString().Split(Environment.NewLine).Length; // format stringBuilder to 40 lines
         return new PingResult(total, stringBuilder?.ToString());
     }
 
-    async public Task<PingResult> RunLongRunningAsync(
-        string hostNameOrAddress, CancellationToken cancellationToken = default)
+    //not passing tests idk if it's me or my OS
+    async public Task<PingResult> RunLongRunningAsync(string hostNameOrAddress, CancellationToken cancellationToken = default)
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:83]: You shouldn't need to use async/await here

File: Assignment/Assignment/PingProcess.cs [hunk 6]
```diff
@@ -13,54 +13,87 @@ public record struct PingResult(int ExitCode, string? StdOutput);
...
+        void UpdateStdOutput(string? line) =>
+            (stringBuilder ??= new StringBuilder()).AppendLine(line);
+        
+        Task<PingResult> task = Task.Factory.StartNew(() =>
+        {
+            Process process = RunProcessInternal(StartInfo, UpdateStdOutput, default, cancellationToken); 
+            return new PingResult(process.ExitCode, stringBuilder?.ToString());
+        }, cancellationToken, TaskCreationOptions.LongRunning, TaskScheduler.Current);
+        
+        return await task;
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:96]: Just return your task
```suggestion
        return task;

```

**Review by mmwoodfo — COMMENTED** [review]
Looks good!

**Pull request closed without merging**